### PR TITLE
Threads for discrete log

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Summary of Changes

Discrete log computation for ElGamal decryption is a very parallelizable operation. I experimented with multithreading the computation and with 4 threads, decryption time seems to decrease by roughly 50~60% by very informal measurements.

I modified the discrete log code so that number of threads can be set as a static constant. I imagine the discrete log computation should be adjusted later for integration, but solving discrete log with 4/8 threads for now does seem to improve test times.

### CODE REVIEW

**Comment by samkim-crypto** [comment]
Yep, okay. Changed the default thread to 1.

File: zk-token-sdk/src/encryption/discrete_log.rs [hunk 1]
```diff
@@ -1,18 +1,16 @@
...
 };
 
-#[allow(dead_code)]
-const TWO15: u64 = 32768;
-#[allow(dead_code)]
-const TWO14: u64 = 16384; // 2^14
 const TWO16: u64 = 65536; // 2^16
-#[allow(dead_code)]
-const TWO18: u64 = 262144; // 2^18
+const NUM_THREADS: usize = 1;
```
Comment (mvines) [inline zk-token-sdk/src/encryption/discrete_log.rs:13]: 🤔 How does the user override this without modifying the source to this crate?

Comment (samkim-crypto) [reply zk-token-sdk/src/encryption/discrete_log.rs:13]: Yes, this has to be adjusted later during integration. For this PR, I thought I would just enable multithreading capability using a constant variable. Or is there a conventional way for users to set the variable?

Comment (samkim-crypto) [reply zk-token-sdk/src/encryption/discrete_log.rs:13]: Oh, I guess the best way is for the decryption function to take in the number of threads as a parameter with default set to 1. Maybe that is what you were pointing out... sorry.

**Comment by samkim-crypto** [comment]
Now, the discrete log construction sets the thread number as 1 by default. This number can be modified using a setter method.

The regular decryption algorithm just decodes the default discrete log instance, but users can customize the thread number as in this [example](https://github.com/solana-labs/solana/pull/23867/files#diff-7d059813b85f149efcc5236cf7f8c29645b18bb9787cdaa05cec84711b06a8a6R599-R601).

File: zk-token-sdk/src/encryption/discrete_log.rs [hunk 2]
```diff
@@ -54,30 +59,83 @@ lazy_static::lazy_static! {
...
             });
+            starting_point -= G;
+        }
+
+        drop(tx);
+        let mut solution = None;
+        for received in rx {
+            solution = Some(received);
+        }
+        solution
```
[the remark this answers is not in the document]
Comment (samkim-crypto) [reply zk-token-sdk/src/encryption/discrete_log.rs:132]: Yes, that is a good point. Not sure if it is the most idiomatic way, but the function now joins the handles before returning.

File: zk-token-sdk/src/encryption/discrete_log.rs [hunk 3]
```diff
@@ -54,30 +59,93 @@ lazy_static::lazy_static! {
             include_bytes!("decode_u32_precomputation_for_G.bincode");
         bincode::deserialize(DECODE_PRECOMPUTATION_FOR_G_BINCODE).unwrap_or_default()
     };
+
+    // pub static ref ITERATOR_G: RistrettoPoint = Scalar::from(NUM_THREADS as u64) * G;
+    // pub static ref THREAD_RANGE_BOUND: usize = (TWO16 as usize) / NUM_THREADS;
```
Comment (mvines) [inline zk-token-sdk/src/encryption/discrete_log.rs:64]:
```suggestion
```

File: zk-token-sdk/src/encryption/discrete_log.rs [hunk 4]
```diff
@@ -54,30 +59,93 @@ lazy_static::lazy_static! {
...
+        Self {
+            generator,
+            target,
+            num_threads: 1,
+            range_bound: TWO16 as usize,
+            step_point: G,
+        }
+    }
+
+    /// Adjusts number of threads ina  discrete log isntance.
```
Comment (mvines) [inline zk-token-sdk/src/encryption/discrete_log.rs:82]:
```suggestion
    /// Adjusts number of threads in a discrete log instance.
```

File: zk-token-sdk/src/encryption/discrete_log.rs [hunk 5]
```diff
@@ -54,30 +59,93 @@ lazy_static::lazy_static! {
...
+    pub fn set_number_threads(&mut self, num_threads: usize) -> Result<(), ProofError> {
```
Comment (mvines) [inline zk-token-sdk/src/encryption/discrete_log.rs:83]: very minor nit: rayon uses `num_threads()` for this ([ref](https://docs.rs/rayon/1.5.1/rayon/struct.ThreadPoolBuilder.html#method.num_threads))
```suggestion
    pub fn num_threads(&mut self, num_threads: usize) -> Result<(), ProofError> {
```

File: zk-token-sdk/src/encryption/discrete_log.rs (hunk 4 above)
Comment (samkim-crypto) [reply zk-token-sdk/src/encryption/discrete_log.rs:82]: Oops, sorry missed ths...

File: zk-token-sdk/src/encryption/discrete_log.rs (hunk 3 above)
Comment (samkim-crypto) [reply zk-token-sdk/src/encryption/discrete_log.rs:64]: Ah missed...

File: zk-token-sdk/src/encryption/discrete_log.rs (hunk 5 above)
Comment (samkim-crypto) [reply zk-token-sdk/src/encryption/discrete_log.rs]: Yes, in addition, maybe it is cleaner to use the builder pattern here as done in raymon, though there is only one parameter here...

Comment (mvines) [reply zk-token-sdk/src/encryption/discrete_log.rs]: I think this is ok as is, no need to reach for the builder pattern in this case IMO

Comment (samkim-crypto) [reply zk-token-sdk/src/encryption/discrete_log.rs]: Okay, sounds good. I'll update the tests with the new method name and it should be good, thanks! 👍

**Pull request merged**